src/libpriv/rpmostree-core.c

```diff
@@ -1564,146 +1564,37 @@ gv_nevra_hash (gconstpointer v)
-                                                   (GDestroyNotify)g_variant_unref);
-    g_autoptr(GPtrArray) packages = dnf_goal_get_packages (goal,
-                                                           DNF_PACKAGE_INFO_UPDATE,
-                                                           DNF_PACKAGE_INFO_DOWNGRADE, -1);
-    for (guint i = 0; i < packages->len; i++)
-      {
-        DnfPackage *pkg = packages->pdata[i];
-        const char *nevra = dnf_package_get_nevra (pkg);
-
-        /* just pick the first pkg */
-        g_autoptr(GPtrArray) old = hy_goal_list_obsoleted_by_package (goal, pkg);
-        g_assert_cmpint (old->len, >, 0);
-        DnfPackage *old_pkg = old->pdata[0];
-
-        /* did we expect this nevra to replace a base pkg? */
-        if (rpmostree_str_ptrarray_contains (replaced_nevras, nevra))
-          g_hash_table_insert (self->pkgs_to_replace, gv_nevra_from_pkg (pkg),
-                                                      gv_nevra_from_pkg (old_pkg));
-        else
-          g_hash_table_insert (forbidden_replacements, g_object_ref (old_pkg),
-                                                       g_object_ref (pkg));
-      }
-
-    if (g_hash_table_size (forbidden_replacements) > 0)
-      {
-        rpmostree_output_message ("Forbidden base package replacements:");
-        GLNX_HASH_TABLE_FOREACH_KV (forbidden_replacements, DnfPackage*, old_pkg,
-                                                            DnfPackage*, new_pkg)
-          {
-            const char *old_name = dnf_package_get_name (old_pkg);
-            const char *new_name = dnf_package_get_name (new_pkg);
-            const char *new_repo = dnf_package_get_reponame (new_pkg);
-            if (g_str_equal (old_name, new_name))
-              rpmostree_output_message ("  %s %s -> %s (%s)", old_name,
-                                        dnf_package_get_evr (old_pkg),
-                                        dnf_package_get_evr (new_pkg), new_repo);
-            else
-              rpmostree_output_message ("  %s -> %s (%s)",
-                                        dnf_package_get_nevra (old_pkg),
-                                        dnf_package_get_nevra (new_pkg), new_repo);
-          }
-        rpmostree_output_message ("This likely means that some of your layered packages "
-                                  "have requirements on newer or older versions of some "
-                                  "base packages. Doing `rpm-ostree cleanup -m` and "
-                                  "`rpm-ostree upgrade` first may help. For "
-                                  "more details, see: "
-                                  "https://github.com/projectatomic/rpm-ostree/issues/415");
-
-        return glnx_throw (error, "Some base packages would be replaced");
-      }
-  }
+  g_assert (!self->pkgs_to_replace);
+  self->pkgs_to_replace = g_hash_table_new_full (gv_nevra_hash, g_variant_equal,
```

Comment:
There are early-returns that could leave this context with a created-but-empty hashtable on errors. Is that ok? If not, what about using a local hash-table and wiring that to `self` just right before the final `return TRUE`.